rts/src/main/java/eta/runtime/io/MemoryManager.java

```diff
@@ -237,4 +242,161 @@ public static long loadStringsUTF8(String[] ss) throws UnsupportedEncodingExcept
     {
         return loadStrings(ss, "UTF-8");
     }
+
+    public static ByteBuffer getBoundedBuffer(long ptr, int offset, int length) {
+        ByteBuffer buf = getBoundedBuffer(ptr);
+        buf.position(buf.position() + offset);
+        buf.limit(buf.position() + length);
+        return buf;
+    }
+
+    public static byte[] getBytes(long ptr, int offset, int length) {
+        ByteBuffer buf = getBoundedBuffer(ptr,offset,length);
+        byte[] b = new byte[buf.remaining()];
+        buf.get(b);
+        return b;
+    }
+    
+    public static ByteBuffer copyByteBuffer( ByteBuffer src, ByteBuffer dest, int n) {
+        return copyByteBuffer(src, 0, dest, 0, n);
+    }
+
+    public static ByteBuffer copyByteBufferDups( ByteBuffer src, ByteBuffer dest, int n) {
+        return copyByteBufferDups(src, 0, dest, 0, n);
+    }
+    
+    public static ByteBuffer copyByteBuffer( ByteBuffer src, int srcOffset
+                                           , ByteBuffer dest, int destOffset
+                                           , int n) {
+        if (src == dest)
+            return dest;
+        src.position(src.position() + srcOffset);
+        src.limit(src.position() + n);
+        dest.position(dest.position() + destOffset);
+        dest.put(src);
+        return dest;
+    }
+
+    public static ByteBuffer copyByteBufferDups( ByteBuffer src, int srcOffset
+                                               , ByteBuffer dest, int destOffset
+                                               , int n) {
+        src = src.duplicate();
+        dest = dest.duplicate();
+        return copyByteBuffer(src, srcOffset, dest, destOffset, n);
+    }
+    
+    public static ByteBuffer copy( long srcAddress, int srcOffset
+                                 , ByteBuffer dest, int destOffset
+                                 , int n) {
+        ByteBuffer src = getBoundedBuffer(srcAddress);
+        return copyByteBuffer(src,srcOffset,dest,destOffset,n);
+    }
+
+    public static ByteBuffer copy( ByteBuffer src, int srcOffset
+                                 , long destAddress, int destOffset
+                                 , int n) {
+        ByteBuffer dest = getBoundedBuffer(destAddress);
+        return copyByteBuffer(src,srcOffset,dest,destOffset,n);
+    }
+
+    
+    public static ByteBuffer copy( long srcAddress, int srcOffset
+                                 , long destAddress, int destOffset
+                                 , int size) {
+        ByteBuffer src  = getBoundedBuffer(srcAddress);
+        ByteBuffer dest = getBoundedBuffer(destAddress);
+        return copyByteBuffer(src,srcOffset,dest,destOffset,size);
+    }
+
+    public static ByteBuffer copy(long srcAddress, long destAddress, int size) {
+        return copy(srcAddress, 0, destAddress, 0, size);
+    }
+
+    public static ByteBuffer set(long address, int val, int size) {
```

Comment:
Maybe this API should take a `byte val` instead of `int val` to clearly convey that only 8 bits will be written? The callee can do the casting.

Reply:
yeah, it is cleaner, i think i copied some of the implementations of eta-hackage